### ISSUE TITLE
Bean validation not picking error translations anymore

### ISSUE THREAD

**Issue by wiecia** [spring-projects/spring-boot#18672]
I have a spring boot application with default bean validation through hibernate validator.
I'm using custom messageSource - which in fact just picks up translations from diferrent places.
On my model classes I'm using custom messages like:
```java
@NotBlank(message = "{validation.error.name.blank}")
```
Using this setup with Spring Boot 2.1.9 everything works fine and my messages are picked up correctly.
With 2.2.0 it shows "{validation.error.name.blank}" as a result. When there is no custom message on annotation: ```@NotBlank``` then it shows default message found in ValidationMessages.properties inside validation jar like "must not be blank".

In my WebMvcConfiguration I'm defining validator bean:
```java
@Override
    public Validator getValidator() {
        LocalValidatorFactoryBean validator = new LocalValidatorFactoryBean();
        validator.setValidationMessageSource(messageSource);
        return validator;
    }
```
I can't find if the is any change according to this in boot or spring framework 5.2

**Comment by snicoll** [comment]
This looks like a duplicate of https://github.com/spring-projects/spring-framework/issues/23839. Let's flag this one on hold and let the framework team figures things out before we start investigating here.

**Comment by snicoll** [comment]
The framework regression has been confirmed so please watch the related issue. If something needs to be done here once the issue has been fixed, we can reopen this issue.

**Comment by wiecia** [comment]
Thank you

**Comment by sbrannen** [comment]
I have reproduced the regression in the following standalone test class.

```java
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.forwardedUrl;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.util.Locale;

import javax.validation.ConstraintViolation;
import javax.validation.Valid;
import javax.validation.constraints.Max;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
import org.springframework.context.MessageSource;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.stereotype.Controller;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.validation.BindingResult;
import org.springframework.validation.FieldError;
import org.springframework.validation.Validator;
import org.springframework.validation.beanvalidation.LocalValidatorFactoryBean;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.context.WebApplicationContext;
import org.springframework.web.servlet.config.annotation.EnableWebMvc;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

@SpringBootTest(webEnvironment = WebEnvironment.MOCK)
class MessageSourceAndValidationTests {

	MockMvc mockMvc;

	@BeforeEach
	void setup(WebApplicationContext wac) {
		this.mockMvc = MockMvcBuilders.webAppContextSetup(wac).build();
	}

	@Test
	void postInvalidForm() throws Exception {
		this.mockMvc.perform(post("/form").param("number", "11"))//
				.andDo(print())//
				.andExpect(status().isOk())//
				.andExpect(forwardedUrl("number_exceeded_10"));
	}

	@Configuration
	// @EnableAutoConfiguration
	@EnableWebMvc
	static class WebConfig implements WebMvcConfigurer {

		@Bean
		MessageSource messageSource() {
			MessageSource messageSource = Mockito.mock(MessageSource.class);
			when(messageSource.getMessage(eq("custom.error.max"), any(), any(Locale.class)))//
					.thenReturn("number_exceeded_{value}");
			return messageSource;
		}

		@Override
		public Validator getValidator() {
			LocalValidatorFactoryBean validator = new LocalValidatorFactoryBean();
			validator.setValidationMessageSource(messageSource());
			return validator;
		}

		@Bean
		FormController formController() {
			return new FormController();
		}

	}

	static class Form {

		@Max(value = 10, message = "{custom.error.max}")
		private int number;

		public int getNumber() {
			return number;
		}

		public void setNumber(int number) {
			this.number = number;
		}
	}

	@Controller
	static class FormController {

		@PostMapping("/form")
		String submit(@Valid Form form, BindingResult errors) {
			if (!errors.hasFieldErrors("number")) {
				throw new IllegalStateException("Declarative validation not applied");
			}

			FieldError fieldError = errors.getFieldError("number");
			if (fieldError.contains(ConstraintViolation.class)) {
				return ((ConstraintViolation<?>) fieldError.unwrap(ConstraintViolation.class)).getMessage();
			}

			return "no_ConstraintViolation_found";
		}

	}

}
```

If you run the test class as-is, it passes.

If you redeclare `WebConfig` as follows...

```java
	@Configuration
	@EnableAutoConfiguration
	// @EnableWebMvc
	static class WebConfig implements WebMvcConfigurer {
        // ...
```

... the test then fails.

In light of that, I suggest that this issue be reopened and investigated on the Boot side.

If there are then steps that need to be taken in Framework, please let us know!

**Comment by sbrannen** [comment]
See https://github.com/spring-projects/spring-framework/commit/47c8d1de7bde543089792cb9517ec540993b113c#diff-410b847334e563dc406577806c2d2027 for the changes in Spring MVC that may be the root cause of this regression.

**Comment by snicoll** [comment]
It's a timing issue and the fact that the `mvcValidator` method in `EnableWebMvcConfiguration` can no longer do its job anymore. To accommodate with various scenarios, we have an [override of this method](https://github.com/spring-projects/spring-boot/blob/2ceffcee14db3f0276bf8897895f833548f25249/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/WebMvcAutoConfiguration.java#L429-L436) that makes sure to expose either a user-configured mvc validator or the auto-configured one if none is present.

When we moved to non-proxied configuration class, the bean methods no longer forces the `mvcValidator` method to be invoked prior to create the `RequestMappingHandlerAdapter`. As the `ValidationAutoConfiguration` runs first, there is a bean that matches the requested predicate and can be used. 

Having said all that, I find it odd that the context does not realize there is going to be two beans of that type in the context ultimately.

**Comment by snicoll** [comment]
With the help of @jhoeller, I've something that works locally by add `@Qualifier("mvcValidator")` on the injection point. We should revisit the existing tests as we have at least one that is supposed to cover that exact use case.
